### ISSUE TITLE
Syncing tickets with different Content-hosts such as Savannah/Github/Bitbucket/local-repo/etc

### ISSUE THREAD

**Issue by heoa** [jeffWelling/ticgit#61]
Now we have a situation:
1. Ticgit has its own Issues under Ticgit -branch
2. Github has its own proprietary Issues, apparently impossible to export similarly to Bitbucket

----> 1&2 sucks. No sync. No co-operation?

**Common problem 1:** tickets between proprietary content-hosts such as Bitbucket.org and Github.org are not synced to TicGit.

**Common problem 2:** TicGitWeb sucks because we don't use it? I even ask whether devs are actually using TicGit?
1. Is it possible to co-operate with Bicbucket.org or Github.com so there is less reinvention of wheels? It is does not look good when Ticgit's own Ticgits are 1-3 years' old at least! 
2. Consult developers in Savannah [2] whether they could add some sort of support for TicGit -issues. They are apparently using some Trac -thing and some free-software so it should be easier to co-operate.

TODO 1: contact Bitbucket.org for API -question, can they read Ticgit -style formatted tickets?
TODO 2: contact Github.com for API -question, can they read Ticgit -style formatted tickets?
TODO 3: if they do not  co-operate, we need find alternative way to the proprietary things.  
TODO 4: is it possible manage TicGit's things through Gitolite so third-parties do not need to care about parsing TicGit's branches? Yes apparently versions/milestones but too much overhead with other things?!

**References**
[1] https://github.com/jeffWelling/ticgit/tree/ticgit
[2] http://savannah.gnu.org/

**Comment by jeffWelling** [comment]
In case you hadn't noticed, a lot of work has been done towards allowing TicGit to sync with third party bug trackers such as Github Issues in a plug-in-able way. Once it is finished, it will ideally be populated with a few simple integrations such as Github Issues, which will provide an example of how other people can write their own plugins so that TicGit can access other bug trackers and synchronize with them as well.

Without such a synchronizing feature, naturally different bug trackers will not collaborate with each other to share tickets.

Problems with TicGitWeb should be broken out into their own tickets instead of being lumped in with others -- I don't think anyone will tell you TicGitWeb is a perfect solution, lots of work is needed, but it's a good proof of concept.  Other web interfaces have been produced for TicGit though they may not be in development anymore.  

The tickets in TicGit are 1-3 years old because I'm the only one who can easily update them.  One solution is for me to update the tickets in the ticgit branch of the TicGit project with the tickets that appear on Github Issues, but a more scalable solution is to write ticket synchronization modules so that pulling and pushing tickets to Github is seamless.  Hence my first paragraph.

I would hold off on contacting anyone about API questions until the synchronization module is finished.   I have contacted Github several times about integration with TicGit and as of yet they have either not commented or said simply 'no'. Github however does provide an API so we can proceed with syncing with Github via the API without need for additional support or integration on their end. Bitbucket may be different, I don't use them and haven't yet checked them out.

I don't know what you mean when you talk about managing TicGit's things through gitolite. What TicGit branches do people have to parse? Is it possible to integrate with gitolite, yes, is it likely in the near future, not by my hand but you're welcome to take up the challenge.

**Comment by jeffWelling** [comment]
The branch where the ticket synchronization work is being done is called 'sync_with_other_repos' if you're interested.  It's not a trivial thing and there is still a problem or two to work out in terms of merge conflict resolution but it shouldn't take too much longer.

**Comment by jeffWelling** [comment]
Could you clarify the title of this ticket a bit more, I'm not sure what you mean or what is supposed to be addressed by this ticket...

**Comment by jeffWelling** [comment]
Based on the current title of the ticket I believe it may be a duplicate of https://github.com/jeffWelling/ticgit/issues/8

**Comment by heoa** [comment]
@jeffWelling Updated and contacted them -- also having discussion with non-commercial providers such as Savannah people that may be more able to help because they are running free software. 

This issue is much more specific. 
1. It is a question about content-host -support with Ticgit. Commercial/proprietary providers may be reluctant to co-operate, not sure waiting for replies.
2. What I want is a dashboard like Gitolite (just ssh/clone the repo to do changes, push to materialize them etc) where you manage permissions and tickets over many repos, below some ideas gathered:

I. tickets -- overhead (managed in each repo or also elsewhere?! How? Submodules to some Gitolite -admin-repo? [1])
II. tickets>milestones -- less overhead (managed in Gitolite)
III. tickets>versions -- less overhead (managed in Gitolite)
1. Non-proprietary providers such as Savannah people may be more able to respond to this syncing problem. We can use other services if not able to get Github/Bitbucket/etc working...

[1] I don't know yet how to sub-modulate the branches but should be easy enough or?

**Comment by jeffWelling** [comment]
Contacted Who?
Help with what?

I'm still not sure what you mean by 1. "Content host support with TicGit"? I don't understand. You think they should host the TicGit branch themselves?...  I don't understand what them helping us would look like, how would that work, I don't think they would want to get down and dirty with our code so that it integrates with their project... frankly that's our problem.

If you want to work on integrating gitolite then I welcome the contributions, but as I've said before, I won't be implementing that feature myself any time soon because I've got too much else on my plate right now.

As long as we are able to read tickets and write tickets (Via API, web interface, third party library, whatever) then we can sync with that bug tracker, we don't _need_ their support in getting syncing working.

**Comment by jeffWelling** [comment]
To be honest I'm still not sure what differentiates this ticket from the one that I linked to above, sorry but can you highlight the differences for me again? Is this about gitolite integration, or is it about ticket synchronization with gitolite and third party bug trackers.... I don't understand..

**Comment by heoa** [comment]
I don't care about individual content-hosts, frankly speaking. Things such as Github are irrelevant, if they do not want to help themselves, they do not want to help themselves -- that is it. I am happy if ticgit has a mode let users to configure their settings to look like settings in other content-hosts [2]. Then after that my plan is to create central place [3] to manage tickets. My refreshed idea is the below with different modes.

Mode 1: allow editing of milestones and versions in Gitolite -style admin repo
Mode 2: allow editing of tickets in Gitolite -style admin repo (different to latter repo?!)
Mode 3: allow third-party -providers to create plugins to TicGit [1](do not hard-code anything, content-hosts or fans are supposed to do this themselves)

Modes 1 and 2 do not need anything from third-parties. Mode 3 is voluntary -- if content-hosts want to offer superb service, perhaps they could help with modes 1 and 2. If the modes can be accomplished, it should not be hard for someone to read the APIs to create specific content-host -support.

**References**
[1] ...after creating a configuration adabtable i.e. solving this ticket [2]
[2] https://github.com/jeffWelling/ticgit/issues/51
[3] https://github.com/jeffWelling/ticgit/issues/57

**Comment by jeffWelling** [comment]
How to do you envision mode 3 working? TicGit itself stores it's tickets in _your_ repository, so unless you've giving access to your repository to some third party, that isn't possible.  What _would_ be possible is for them to integrate with, say, Github Issues or some other 'fourth' party bug tracker.  They push their tickets to that fourth party, your ticgit pulls from that fourth party, and vice versa, there you have it.

Besides that, I don't know how else it would work. 

If you host your repository on Github or Gitorious, then a third party could be able to read your tickets by scanning your TicGit branch, but in that scenario they wouldn't have write access.

**Comment by heoa** [comment]
For modes I and II, read this [1] (editing of TicGits goes simply fetching the TicGits, editing and pushing back). For mode I, read this [2]. For pluggability -- well we cannot do much. By far the easiest for us would be the solution would X, instead of Y or Z.

X. When user starts creating tickets in a content-host, the content-host ask whether s/he wants TicGit -compability (i.e. creating a branch for tickets) -- unless no co-operation between major players such as Bitbucket.org, Github.com, Savannah/etc, we are going to see incompatible issues... (still young thing so most producers probably careful here, looking how things evolve -- and they are probably much more interested to lock users...)

Y. Content-host: read their probably evolving APIs, waste your time. Work "for free" to fix compability issues and adding support. (bad, easiest would be if the parties tried to make some sort of aggreement or even allowed some sort of exportation of tickets)

Z. Compete against the Content-hosts and create TicGit -compatible services with other things such as [3]

**References**
[1] https://github.com/jeffWelling/ticgit/issues/57
[2] https://github.com/jeffWelling/ticgit/issues/51
[3] http://stackoverflow.com/questions/11251941/how-can-i-see-github-style-things-such-as-punchcards-and-timeline-of-git-repo

**Comment by jeffWelling** [comment]
This is sounding more and more like the duplicate linked to above. I don't see any differences between that ticket and this one...

**Comment by heoa** [comment]
Bitbucket.org said '[not] this time' but they welcomed contributions using the API [1]. I suggest to forget and move on.

[1] https://confluence.atlassian.com/display/BITBUCKET/Using+the+bitbucket+REST+APIs

**Comment by jeffWelling** [comment]
Did you check to see if we even need any support additional to the API? Much of the time the API is sufficient and we don't need them to create any kind of 'ticket export' system, the API does just fine.  I haven't checked the BitBucket API, but have you?

**Comment by heoa** [comment]
## Bitbucket.org

Step 0: If doing things on your own computer, one could easily do things with a few bash scripts. It is pretty easy:

$ curl https://api.bitbucket.org/1.0/repositories/sarahmaddox/sarahmaddox/issues/ --data "title=Issue%20Title&content=Issue%20Content" --user user_name:password"

Step 1: For automation, use OAuth 1.0 [1] authentication
Step 2: Create issue [2] etc

[1] https://confluence.atlassian.com/display/BITBUCKET/OAuth+on+Bitbucket
[2] https://confluence.atlassian.com/display/BITBUCKET/Issues#Issues-CreatinganIssue
